src/python/pants/engine/README.md

```diff
@@ -49,44 +50,44 @@ then the `@rule` will run exactly once, and the value that it produces will be a
+`Product` given a single `int` argument.
 
 ```python
-@rule(str, [int])
-def int_to_str(an_int):
-  return str(an_int)
+@rule
+def int_to_str(value: int) -> str:
+  return str(value)
 ```
 
-The first argument to the `@rule` decorator is the `Product` (ie, return) type for the `@rule`. The
-second argument is a list of "parameter selectors" that declare the types of the input parameters for
-the `@rule`. In this case, because the `Product` type is `str` and there is one parameter
-selector (for `int`), this `@rule` represents a conversion from `int` to `str`, with no other inputs.
+Here the only requirements for the `@rule` are:
+1. The decorated function is a free, top-level function; ie: methods cannot be `@rule`s.
+2. The decorated function must be fully type annotated.
+3. As mentioned above, the types of the function parameters must be hashable and have useful equality.
 
-When the engine encounters this `@rule` while compiling the rule graph for `str`-producing-`@rules`,
+When the engine encounters this `@rule` while compiling the rule graph for `str`-producing-`@rule`s,
 it will next go hunting for the dependency `@rule` that can produce an `int` using the fewest number
 of `Params`. For example, if there was an `@rule` that could produce an `int` without consuming any
-`Params` at all (ie, a singleton), then that `@rule` would always be chosen first. If all `@rules` to
-produce `int`s required at least one `Param`, then the engine would next see whether the input `Params`
-contained an `int`, or whether there were any `@rules` that required only one `Param`, then two
-`Params`, and so on.
+`Params` at all (ie, a singleton), then that `@rule` would always be chosen first. If all `@rule`s
+to produce `int`s required at least one `Param`, then the engine would next see whether the input
+`Params` contained an `int`, or whether there were any `@rule`s that required only one `Param`, then
+two `Params`, and so on.
 
-In cases where this search detects any ambiguity (generally because there are two or more `@rules` that
-can provide the same product with the same number of parameters), rule graph compilation will fail with
-a useful error message.
+In cases where this search detects any ambiguity (generally because there are two or more `@rule`s
+that can provide the same product with the same number of parameters), rule graph compilation will
+fail with a useful error message.
 
 ### Datatypes
 
-In practical use, builtin types like `str` or `int` do not provide enough information to disambiguate
-between various types of data in `@rule` signatures, so declaring small `datatype` definitions to
-provide a unique and descriptive type is highly recommended:
+In practical use, builtin types like `str` or `int` do not provide enough information to
+disambiguate between various types of data in `@rule` signatures, so declaring small `datatype` 
+definitions to provide a unique and descriptive type is highly recommended:
```

Comment:
(thanks for updating the paragraph filling in this doc!)

Reply:
Ditto

Reply:
I did nothing of the sort that I can tell, I just line-wrapped.